### ISSUE TITLE
making Spotify Podcasts accessible to your favourite Podcast client

### ISSUE THREAD

**Issue by alexanderadam** [SamTV12345/PodFetch#143]
### Clear and concise description of the problem

Just think about it: what if PodFetch would have support for other 'proprietary Providers' like Spotify?
Platforms like Spotify [claim to have Podcasts](https://open.spotify.com/genre/podcasts-web) yet the _original_ idea of Podcasts were standardised RSS feeds and sadly not every 'Podcast' on Spotify has an RSS feed. Often their Podcasts are only accessible within their own client or ['hacks' like this one](https://github.com/timdorr/spotifeed).

There are _some_ feeds that are accessible and some that aren't.
So I'm thinking that global or user specific provider credentials (i.e. for Spotify in this case) could be added and PodFetch would handle it like a regular RSS feed and provide a proper mirrored feed for it, too.

This would also help in cases where [some Podcasts are region restricted](https://community.spotify.com/t5/Other-Podcasts-Partners-etc/Does-Spotify-podcasts-have-regional-restrictions/td-p/5338369) and there fore liberate knowledge for people on different levels. Furthermore it might [keep the web open](https://www.wired.co.uk/article/spotify-exclusive-podcasts) ([1](https://mattstoller.substack.com/p/on-the-spotify-joe-rogan-deal-and), [2](https://mattstoller.substack.com/p/will-spotify-ruin-podcasting), [3](https://www.techdirt.com/2020/07/17/we-shouldnt-call-michelle-obamas-joe-rogans-proprietary-exclusive-audio-spotify-podcast-any-more/)).

**PS:**  It would probably also make sense to integrate Podcast results from Spotify search
**PPS:** This is surely a tough feature but I thought the same [with my last idea](https://github.com/SamTV12345/PodFetch/issues/41#issue-1652706823) and you mastered this feature super quickly so I'm not sure you're considering things a difficult when I think they are. :wink: 

### Suggested solution

Apparently there are crates like [rspotify](https://docs.rs/rspotify/latest/rspotify/), [spotify_api](https://docs.rs/spotify-api/0.1.2/spotify_api/), [spotify](https://docs.rs/spotify/0.8.0/spotify/) and [many others](https://docs.rs/releases/search?query=spotify+api) that abstract the provider specific API calls.

### Alternative

There were projects like [this one](https://github.com/timdorr/spotifeed)

### Validations

- [X] Follow our [Code of Conduct](https://github.com/SamTV12345/.github/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guide](https://github.com/SamTV12345/.github/blob/main/CONTRIBUTING.md).
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

**Comment by SamTV12345** [comment]
There is a mention in the documentation. I have no idea if this is legal as I would circumvent their security mechanisms:

> The technical details: Spotify's anon-podcast.scdn.co server has been
>         discontinued. This served as an unauthenticated proxy server for podcast
>         episode audio. They have switched to streaming directly from the
>         podcast's source server. For non-exclusive podcasts, that is simply
>         whatever is in the RSS feed (in which case, you can access those
>         podcasts directly without Spotifeed's help). For exclusive podcasts,
>         Spotify is using their audio-fa.scdn.co server, which appears to be
>         using a chunked MP4 stream that is not usable by general podcast
>         clients. That means Spotifeed would have to re-host the podcast audio
>         files to make those downloadable, which is not feasible or legal.

**Comment by alexanderadam** [comment]
I don't have any clue about the legal state of it either.
Should I close this issue?

**Comment by SamTV12345** [comment]
You could leave it open for now.  Maybe someone has an idea on how to do that without having the Spotify legal team sueing you.

**Comment by alexanderadam** [comment]
How about having this feature disabled behind an ENV variable that makes it super obvious that the contributors won't take any responsibility about its usage.

Something like

```
ACTIVATE_SPOTIFY_SYNC_WITHOUT_ANY_RESPONSIBILITY_OF_THE_CONTRIBUTORS_AND_MAINTAINERS=yes
```

And maybe writing something like this in the license might make sense as well I guess.

Some Bittorrent client maintainers have [these lines included](https://github.com/transmission/transmission/blob/main/COPYING#L10-L12):

> This program is distributed in the hope that it will be useful,
> but WITHOUT ANY WARRANTY; without even the implied warranty of
> MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

**Comment by SamTV12345** [comment]
Something like this should be appropriate. But it can take some time.

**Comment by alexanderadam** [comment]
> Something like this should be appropriate. But it can take some time.

I fully understand that.
I'm still not sure how you were able to implement the gPodder API this fast 😄

**Comment by SamTV12345** [comment]
Spotify is the ridiculous company I have ever thought they would be. They send every 5-10 seconds rolling keys that encrypt the next part of the chunk. In order to decrypt an hour long play you need to listen to the full podcast. Doing this automatic would trigger the Google Captcha on their side as I would need to run this in a real browser to get the keys. You don't even have an audio tag or something else that streams from an url. Everything is handled via GraphQL resulting in so many requests and traffic that I'm astonished that this company still exists. 

TLDR: I have no idea how I would start with this task. Spotify seems to have won on that and there is no possibility to download a podcast.

**Comment by alexanderadam** [comment]
Today I learned about [this](https://github.com/Rafiuth/Soggfy#how-it-works) and [this](https://web.archive.org/web/20200303145624/https://github.com/meik97/XSpotify).

This is indeed ridiculous.
Well, I guess I should close this then.

Thank you that you took the time to look into this though. I'm sure that it would've been a huge selling point for many.

**Comment by alexanderadam** [comment]
I guess [spotifyd](https://github.com/Spotifyd/spotifyd) is doing exactly this, right?

**Comment by SamTV12345** [comment]
I'm not sure if that works either. If I read correctly running this is similar to the desktop app. I'm not sure if this performs any download and you need a premium subscription.

**Comment by alexanderadam** [comment]
Btw I'm not sure whether I should keep this issue closed or open.

The [Podgrab issue](https://github.com/akhilrex/podgrab/issues/210) is open too.

Maybe someone who sees this ticket as open, has further input.

What do you think @SamTV12345 ?

**Comment by SamTV12345** [comment]
Yes. Let's reopen this. There is also a new crate available: https://github.com/librespot-org/librespot

**Comment by hugolinhares** [comment]
Hey, I have found this project that might help: [https://github.com/Yetangitu/Spodcast/issues](url)

**Comment by SamTV12345** [comment]
Hmm. How would I use that? It's written in Python so I would have to invoke it via CLI. Maybe a rewrite in Rust as an actual library would work.
